### PR TITLE
ARTEMIS-1205: AMQP Shared Durable Subscriber incorrect behaviour.

### PR DESCRIPTION
**Pull request by michaelandrepearce**
Add test case, to prove the issue, and then obviously ensure it works, post fix.
Apply changes in logic of createQueueName to handle global better and fix the behaviour.

### CODE REVIEW

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/ProtonServerSenderContext.java [hunk 1]
```diff
@@ -738,17 +738,14 @@ private static String createQueueName(String clientId,
                                          boolean shared,
                                          boolean global,
                                          boolean isVolatile) {
-      String queue = clientId == null || clientId.isEmpty() ? pubId : clientId + "." + pubId;
+      String queue = clientId == null || clientId.isEmpty() || global ? pubId : clientId + "." + pubId;
```
Comment (clebertsuconic) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/ProtonServerSenderContext.java]: 👍 ...
I will run the testsuite and merge this

Comment (michaelandrepearce) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/ProtonServerSenderContext.java]: wait one second I'm just sorting our SharedConsumer also :) , just pushing :)

Comment (michaelandrepearce) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/ProtonServerSenderContext.java]: pushed. :)

Comment (michaelandrepearce) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/ProtonServerSenderContext.java]: And build is green :)

**Comment by michaelandrepearce** [comment]
@clebertsuconic def no worries, entirely makes sense. Was just updating to let you know everything is ready from my POV.

**Comment by clebertsuconic** [comment]
This introduced a few failures on this test: 

org.apache.activemq.artemis.tests.integration.amqp.ClientDefinedMultiConsumerTest

**Comment by clebertsuconic** [comment]
Normal failures... because of the queue names have changed.

**Comment by michaelandrepearce** [comment]
@clebertsuconic should be fixed now.

**Comment by clebertsuconic** [comment]
@michaelandrepearce  I'm thinking about reverting this.. this will cause compatibility issues...

Can we fix only the clientID part...


if you really want a new queue naming convention to match core and AMQP, we will need a property somewhere to change the semantic

**Comment by michaelandrepearce** [comment]
But amqp jms shared consumer is functionally broken without this

**Comment by michaelandrepearce** [comment]
As in it is not breaking anything

**Comment by michaelandrepearce** [comment]
But on fixing as we won't be breaking anything we are taking the opportunity to align

**Comment by clebertsuconic** [comment]
That is what I'm trying to access.  If it's just one case broken I would fix only this case.  But if it's completely broken then yes we use the opportunity to fix the whole thing.

**Comment by michaelandrepearce** [comment]
Yes it was compleatly broken, that what I added the tests for. If you revert and took just the amqp tests I added you'll note they'd fail

**Comment by clebertsuconic** [comment]
@michaelandrepearce that was my impression as well.. but then @gemmellr told me that it was only the case of the clientID that was broken.. according fixing this would break everything else from the POV of compatibility.

So, I"m trying to figure out what is the case now 👍

**Comment by gemmellr** [comment]
The no-ClientID shared-subscription behaviour was indeed broken for AMQP JMS clients. Fixing that issue requires ensuring the container-id is not included in their backing queue name by the broker. Changing those queue names in that way wouldnt break existing AMQP JMS no-ClientID shared-subscription usage because they were already broken. All good.

Introducing the alignment changes in addition to fixing the container-id bug does however introduce other issues that were not previously present for AMQP clients (see the JIRA for some examples), and additionally the latest change to fix alignment of "." handling could actually break existing working subscriptions upon upgrade. So I'd say the question is, is the alignment important enough to introduce those issues for AMQP clients and also potentially break some existing subscriptions? I'm not sure it is, but even if so I'm also not sure I'd do it in a minor version, and at a major version change I would probably instead consider changing the Core side.

I'm heading out shortly, so I won't be around for further replies this evening.

**Comment by michaelandrepearce** [comment]
How does it the alignment changes are in the block of code in the if(shared) section which only applies to shared subscribers the part in point that's broken

**Comment by michaelandrepearce** [comment]
Agreed the "." Second pr could

**Comment by michaelandrepearce** [comment]
Ok what I propose is, we keep the changes that were in this pr that only affect shared subscribers (broken) and we rework the alignment of the escaped "." In the other to actually maybe in the Artemis client not to escape (but make it a feature flag change)

Is this reasonable?

**Comment by michaelandrepearce** [comment]
For us the alignment is fairly critical, as we will have shared core and amqp env

**Comment by michaelandrepearce** [comment]
How does this sound to you clebert?

**Comment by clebertsuconic** [comment]
as long as the default is the current one, and the flag only applies to the client.. than I'm fine...

I would prefer leaving as is.. just one less moving part.. but if that's critical for you I'm fine.. as Long as it doesn't get too bloated in the code. (hard to figure out)...


So, if you create a connection factory property.. it's probably ok.. defaulted to what there is now.

**Comment by michaelandrepearce** [comment]
Sure will do this.

**Comment by clebertsuconic** [comment]
```
For us the alignment is fairly critical, as we will have shared core and amqp env
```

Are you sure you need this in production? or that's a testcase only.


anyways.. I"m out for 2 weeks starting tomorrow... I'm trying to get a shock treatment to get away from the iphone as much as I can. :)

**Comment by michaelandrepearce** [comment]
https://github.com/apache/activemq-artemis/pull/1326 raised this to revert the code to just what we agreed. It is critical for us as we have some apps/clients that will be sharing subscriptions using amqp and core client, like wise we will be configuring up front the queues for multicast, and it makes it very tricky/almost asking for a disaster for the operations guys to have two differing formats.

**Comment by michaelandrepearce** [comment]
I will most probably won't get the chance tonight to do the other bit, as I'm already beer in hand readying myself for a long weekend in the sun too :)  

If you could maybe tee up with @jbertram to look at that for me when i raise next week if you'll be away.

**Comment by michaelandrepearce** [comment]
After looking to revert this. I came back across another reason why we had to do this.

There still is a fundamental problem with leaving :global in as it stands (so this bit to make it meet spec i can't remove).

We should note, as per JMS Spec:http://docs.oracle.com/javaee/7/api/javax/jms/Session.html#unsubscribe-java.lang.String-

"
A shared durable subscription and an unshared durable subscription may not have the same name and client identifier (if set). If an unshared durable subscription already exists with the same name and client identifier (if set) then a JMSException is thrown.
"

This means the current behaviour in AMQP here is wrong, as shared and unshared with the same subscription name are not making the same queue as such above is invalid.

e.g. if clientId = myClient;

createSharedDurableConsumer(myTopic, "mySub")
will create myClient.mySub:global

createDurableConsumer(myTopic, "mySub")
will create myClient.mySub

Essentially this is why we need to remove the :global part still in the shared part, otherwise this is breaking spec. The other option would be to add :global (or a rename of :global to :durable) to the durable consumer, but that would break things historically.

As removing :global in shared wouldn't break anything already historically, as SharedDurable is not in a working state. this is why we do have to do this.

**Comment by michaelandrepearce** [comment]
Re concerns on the overlap of a shared consumer queue and another queue space, it is already possible to make a jms queue named myClient.mySub:global as such we already have collision issue since 2.0. This change does not increase this risk/problem. I think it is something that users will have to be aware of and just configure/ensure they don't do things that risk that (as it is already an issue!)

**Comment by gemmellr** [comment]
Given that all the queues exist in the same ultimate namespace, there will obviously always be potential for collisions unless 'regular queues' had some name mangling applied too (which is undesirable, and was removedin 2.x because of it). I don't dispute this, I just think it is a question of severity and whether one subscription type by default will/could easily collide with another as well as entirely unrelated things such as regular queues with simple names.

Someone with a queue called "foo", and a topic called "bar" and a client who makes a no-ClientID subscription named "foo", should be allowed to. With the existing format (minus the container-id bug) they would be able to on the AMQP side. With the aligned one they will not as the queue and subscription names would be directly used the same way without attempt to avoid collision in use of these essentially unrelated values.

The addition of a ":global" suffix was not incorrect, the implementation simply had a related bug in it. The quoted JMS spec excerpt (I assume is from createSharedDurableConsumer, not the linked unsubscribe) and example are not actually applicable to it, as the ":global" suffix only applies for shared subscriptions on connections without a ClientID, which cant have non-shared durable subscriptions (i.e they need a ClientID to be set). The ":global" suffix exists precisely to separate those 'global sub backing' queue names from being named directly the same way as regular queues and other sub types when presented with simple name "foo". If you go to the effort of naming all your regular queues "foo:global" etc then I agree you can still cause the same collisions, but I think at that point its actually reasonble to say there is a naming convention and you need to take it into account, as its far less likely to be hit and the default for every simple name "foo" was not for them to collide.

The ":shared-volatile" and ":global" suffixes also help prevent a durable [shared] subscription backing queue name from easily colliding with a non-durable one, or with those from connections with ClientID's. Removing those with the alignment, that then becomes somewhat more likely, unless the "." escaping is done, which as discussed could break existing subscriptions. Obviously, use of regular queues could again be made to collide names with any naming convention used here, but again for me the key bit is typically only with some deliberate effort.

The issue noted with unsubscribe (which I was planning to test+fix after this JIRA was complete) will most likely be that the broker is not looking in the correct place for the "global" flag as it applies during unsubscribe. This is unfortunately in a different place than it is the rest of the time due to how the complicated AMQP<->JMS mapping must work for shared subs. Typically the flag is added as an AMQP 'source capability', but during unsubscribe the is no 'source' information present so instead it gets passed as a 'link desired capability'. The change to add the below would actually work for the JMS client, but would typically fail/be-incorrect for any non-JMS clients using the mechanism. Whats needed is inspecting the link capabilities and passing the correct global value into the naming method.
```
+      if (pubId.endsWith("|global")) {
+         global = true;
+      }
```

**Comment by gemmellr** [comment]
I seem to be the only one that considers the direct collision of simple 'regular queue' names and 'no-Client shared subscription backing queue' names to be an issue. If everyone else is happy I'm not standing in the way. I just feel the need to point it out as I think it is a fairly annoying usability issue that shouldnt exist. Personally I'd change both the Core and AMQP sides together in a later 3.0.0 release.

If the alignment is done then I'd possibly go ahead and align the AMQP "." handling behaviour as well, even though it could break some existing subscriptions (hopefully there aren't [m]any affected), given other alignment issues would actually be lessened by it and if the aim is alignment then it would make sense they actually do align.

**Comment by tabish121** [comment]
I'm in agreement with Robbie on this, I think the ease of which the subscription names and "regular queue" names can collide should be avoided.  It's much easier to tell people (document) how to avoid this collision if we have in place some means of at least making it something you couldn't do without putting some deliberate effort into it or have some amazingly bad luck on your side.  

Lining up Core and AMQP mapping names for subscriptions seems like a good thing to add to a 3.0 task list for changes that require a major version bump.  In the meantime if some configuration switch is added to control using one naming scheme or another it ought to be added as an already deprecated sort of thing that is going away in the next major version.

**Comment by michaelandrepearce** [comment]
Hi, 
I have raised 2 PR's 

one is to continue/complete on the above work and align using CORE client side configurable, based on @clebertsuconic comment that he's happy to have that client side in core. 
https://github.com/apache/activemq-artemis/pull/1328

one is to move this logic and to align in the AMQP protocol server side configurable. Based on the last two comments that seems having it configurable so that its configurable if to break existing behaviour, but having the fixes or chose to remain as is (default)
https://github.com/apache/activemq-artemis/pull/1337

The idea being we chose the more preferable solution out of the two.

I am guessing @gemmellr the second (1337) is your more preferable one. Is every one agreed?

**Pull request merged**